### PR TITLE
Change forEach to for...of

### PR DESCRIPTION
**Pull request by TheJetOU**
~This is a WIP, because it breaks a lot of things (specifically the `forEach`'s with multiple params), so I will need to test each one of them and fix accordingly.~

### CODE REVIEW

**Comment by Zarel** [comment]
This pull request **introduces 1 alert** when merging c766ce0e5a203bc45a5b23b8068dc09116779d02 into 295f7c6c9f3620829801813daae401a97a48e1f3 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-d1a2727983bfb972f967ea8a5055c7f2ce585259)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: dev-tools/users-utils.js [hunk 1]
```diff
@@ -46,9 +46,9 @@ class Worker extends EventEmitter {
 			throw new Error(`Attempted to disconnect nonexistent socket ${socketid}`);
 		}
 		this.sockets.delete(socketid);
-		this.rooms.forEach((room, roomid) => {
+		for (const [room] of this.rooms) {
```
Comment (Zarel) [inline dev-tools/users-utils.js]:
```suggestion
		for (const [roomid, room] of this.rooms) {
```

File: server/chat-commands.js [hunk 2]
```diff
@@ -2468,9 +2468,9 @@ const commands = {
 		target = this.canHTML(target);
 		if (!target) return;
 
-		Users.users.forEach(u => {
+		for (const u of Users.users) {
```
Comment (Zarel) [inline server/chat-commands.js]:
```suggestion
		for (const u of Users.users.values()) {
```

File: server/chat-commands.js [hunk 3]
```diff
@@ -3401,7 +3401,9 @@ const commands = {
 			return this.errorReply("Wait for /updateserver to finish before using /kill.");
 		}
 
-		Sockets.workers.forEach(worker => worker.kill());
+		for (const worker of Sockets.workers) {
```
Comment (Zarel) [inline server/chat-commands.js]:
```suggestion
		for (const worker of Sockets.workers.values()) {
```

File: server/chat-plugins/info.js [hunk 4]
```diff
@@ -328,35 +328,35 @@ const commands = {
...
+			}
 			if (results.length > 100 && !isAll) {
 				return this.sendReply(`More than 100 users match the specified IP range. Use /ipsearchall to retrieve the full list.`);
 			}
 		} else if (ip.slice(-1) === '*') {
 			// IP range
 			this.sendReply(`Users in IP range ${ip}${targetRoom ? ` in the room ${targetRoom.title}` : ``}:`);
 			ip = ip.slice(0, -1);
-			Users.users.forEach(curUser => {
+			for (const curUser of Users.users) {
```
Comment (Zarel) [inline server/chat-plugins/info.js]:
```suggestion
			for (const curUser of Users.users.values()) {
```

**Comment by Zarel** [comment]
Okay, at this point, I feel like I shouldn't keep doing this review.

The important thing you need to understand here is the difference between a `Map` and an `Array`, and how to iterate them.

Arrays and Sets work like this:

`array.forEach(value => {...});`

=> `for (const value of array) {...}`

Maps work like this:

`map.forEach(value => {...});`

=> `for (const value of map.values()) {...}`

`map.forEach((value, key) => {...});`

=> `for (const [key, value] of map) {...}`

File: server/chat-commands.js [hunk 5]
```diff
@@ -2482,9 +2482,9 @@ const commands = {
 		target = this.canHTML(target);
 		if (!target) return;
 
-		Rooms.rooms.forEach((curRoom, id) => {
-			if (id !== 'global' && curRoom.type !== 'battle') curRoom.addRaw(`<div class="broadcast-blue"><b>${target}</b></div>`).update();
-		});
+		for (const curRoom of Rooms.rooms.values()) {
+			if (curRoom.id !== 'global' && curRoom.type !== 'battle') curRoom.addRaw(`<div class="broadcast-blue"><b>${target}</b></div>`).update();
```
Comment (Zarel) [inline server/chat-commands.js]:
```suggestion
			if (curRoom.id !== 'global' && curRoom.type !== 'battle') {
				curRoom.addRaw(`<div class="broadcast-blue"><b>${target}</b></div>`).update();
			}
```

File: server/chat-plugins/trivia.js [hunk 6]
```diff
@@ -923,7 +923,7 @@ class TimerModeTrivia extends Trivia {
 		}
 
 		let rowAdded = false;
-		innerBuffer.forEach((players, pointValue) => {
+		for (let [pointValue, players] of innerBuffer) {
 			if (!players.length) return false;
```
Comment (Zarel) [inline server/chat-plugins/trivia.js]:
```suggestion
			if (!players.length) continue;
```

`return` becomes `continue` when refactoring `.forEach` to `for...of`

File: server/chat-plugins/info.js [hunk 7]
```diff
@@ -328,35 +328,35 @@ const commands = {
...
-			});
+			}
 			if (results.length > 100 && !isAll) {
 				return this.sendReply(`More than 100 users match the specified IP range. Use /ipsearchall to retrieve the full list.`);
 			}
 		} else {
 			this.sendReply(`Users with IP ${ip}${targetRoom ? ` in the room ${targetRoom.title}` : ``}:`);
-			Users.users.forEach(curUser => {
+			for (const curUser of Users.users.values()) {
 				if (curUser.latestIp !== ip) return;
```
Comment (Zarel) [inline server/chat-plugins/info.js]:
```suggestion
				if (curUser.latestIp !== ip) continue;
```

File: server/chat-plugins/info.js [hunk 8]
```diff
@@ -328,35 +328,35 @@ const commands = {
...
 				if (targetRoom && !curUser.inRooms.has(targetRoom.id)) return;
```
Comment (Zarel) [inline server/chat-plugins/info.js]:
```suggestion
				if (targetRoom && !curUser.inRooms.has(targetRoom.id)) continue;
```

File: server/chat-plugins/info.js [hunk 9]
```diff
@@ -328,35 +328,35 @@ const commands = {
...
 			if (results.length > 100 && !isAll) {
 				return this.sendReply(`More than 100 users match the specified IP range. Use /ipsearchall to retrieve the full list.`);
 			}
 		} else if (ip.slice(-1) === '*') {
 			// IP range
 			this.sendReply(`Users in IP range ${ip}${targetRoom ? ` in the room ${targetRoom.title}` : ``}:`);
 			ip = ip.slice(0, -1);
-			Users.users.forEach(curUser => {
+			for (const curUser of Users.users.values()) {
 				if (results.length > 100 && !isAll) return;
```
Comment (Zarel) [inline server/chat-plugins/info.js]:
```suggestion
				if (results.length > 100 && !isAll) continue;
```

File: server/chat-plugins/info.js [hunk 10]
```diff
@@ -328,35 +328,35 @@ const commands = {
...
 				if (!curUser.latestIp.startsWith(ip)) return;
```
Comment (Zarel) [inline server/chat-plugins/info.js]:
```suggestion
				if (!curUser.latestIp.startsWith(ip)) continue;
```

File: server/chat-plugins/info.js [hunk 11]
```diff
@@ -328,35 +328,35 @@ const commands = {
...
 				if (targetRoom && !curUser.inRooms.has(targetRoom.id)) return;
```
Comment (Zarel) [inline server/chat-plugins/info.js]:
```suggestion
				if (targetRoom && !curUser.inRooms.has(targetRoom.id)) continue;
```

File: server/chat-plugins/info.js [hunk 12]
```diff
@@ -328,35 +328,35 @@ const commands = {
 		if (/[a-z]/.test(ip)) {
 			// host
 			this.sendReply(`Users with host ${ip}${targetRoom ? ` in the room ${targetRoom.title}` : ``}:`);
-			Users.users.forEach(curUser => {
+			for (const curUser of Users.users.values()) {
 				if (results.length > 100 && !isAll) return;
```
Comment (Zarel) [inline server/chat-plugins/info.js]:
```suggestion
				if (results.length > 100 && !isAll) continue;
```

File: server/chat-plugins/info.js [hunk 13]
```diff
@@ -328,35 +328,35 @@ const commands = {
...
 				if (!curUser.latestHost || !curUser.latestHost.endsWith(ip)) return;
```
Comment (Zarel) [inline server/chat-plugins/info.js]:
```suggestion
				if (!curUser.latestHost || !curUser.latestHost.endsWith(ip)) continue;
```

File: server/chat-plugins/info.js [hunk 14]
```diff
@@ -328,35 +328,35 @@ const commands = {
...
 				if (targetRoom && !curUser.inRooms.has(targetRoom.id)) return;
```
Comment (Zarel) [inline server/chat-plugins/info.js]:
```suggestion
				if (targetRoom && !curUser.inRooms.has(targetRoom.id)) continue;
```

File: server/chat-commands.js [hunk 15]
```diff
@@ -2898,41 +2898,41 @@ const commands = {
...
 			}
 		}
 
 		let soonExpiring = (cmd === 'expiringblacklists' || cmd === 'expiringbls');
 		let buf = Chat.html`Blacklist for ${room.title}${soonExpiring ? ` (expiring within 3 months)` : ''}:<br />`;
 
-		blMap.forEach((data, userid) => {
+		for (const [userid, data] of blMap) {
 			const [expireTime, ...alts] = data;
 			if (soonExpiring && expireTime > Date.now() + SOON_EXPIRING_TIME) return;
```
Comment (Zarel) [inline server/chat-commands.js]:
```suggestion
			if (soonExpiring && expireTime > Date.now() + SOON_EXPIRING_TIME) continue;
```

File: server/punishments.js [hunk 16]
```diff
@@ -66,11 +66,11 @@ class PunishmentMap extends Map {
 	/**
 	 * @param {(punishment: Punishment, id: string, map: PunishmentMap) => void} callback
 	 */
-	forEach(callback) {
-		super.forEach((punishment, k) => {
+	forOf(callback) {
```
Comment (Zarel) [inline server/punishments.js]: lol

This is actually something to watch out for; you might even want to implement this as an iterator.

But if you don't, I'd just keep this unrenamed.

File: server/punishments.js [hunk 17]
```diff
@@ -66,11 +66,11 @@ class PunishmentMap extends Map {
...
+		for (const [k, punishment] of super.entries()) {
 			if (Date.now() < punishment[2]) return callback(punishment, k, this);
```
Comment (Zarel) [inline server/punishments.js]:
```suggestion
			if (Date.now() < punishment[2]) {
				callback(punishment, k, this);
				continue;
			}
```

File: server/punishments.js [hunk 18]
```diff
@@ -135,13 +135,13 @@ class NestedPunishmentMap extends Map {
 	/**
 	 * @param {(punishment: Punishment, roomid: string, userid: string) => void} callback
 	 */
-	nestedForEach(callback) {
-		this.forEach((subMap, k1) => {
-			subMap.forEach((punishment, k2) => {
+	nestedForOf(callback) {
```
Comment (Zarel) [inline server/punishments.js]: Ditto, leave this unrenamed.

File: server/tournaments/generator-elimination.js [hunk 19]
```diff
@@ -67,16 +67,16 @@ class Elimination {
 	}
 	getUsers(remaining) {
 		let users = [];
-		this.users.forEach((value, key) => {
+		for (const [key, value] of this.users) {
 			if (remaining && (value.isEliminated || value.isDisqualified)) return;
```
Comment (Zarel) [inline server/tournaments/generator-elimination.js]:
```suggestion
			if (remaining && (value.isEliminated || value.isDisqualified)) continue;
```

File: server/tournaments/generator-elimination.js [hunk 20]
```diff
@@ -67,16 +67,16 @@ class Elimination {
...
 		return users;
 	}
 
 
 	generateBracket() {
-		Dex.shuffle(this.getUsers()).forEach(user => {
+		for (const user of Dex.shuffle(this.getUsers())) {
 			if (!this.tree) {
 				this.tree = {
 					tree: new TreeNode(null, {user: user}),
```
Comment (Zarel) [inline server/tournaments/generator-elimination.js]: There's a return around five lines below this line that needs to be converted to `continue`

File: server/tournaments/generator-round-robin.js [hunk 21]
```diff
@@ -113,24 +113,24 @@ class RoundRobin {
 
 		let userIndex = this.users.indexOf(user);
 
-		this.matches[userIndex].forEach((match, col) => {
+		for (const [col, match] of this.matches[userIndex].entries()) {
 			if (!match || match.state !== 'available') return;
```
Comment (Zarel) [inline server/tournaments/generator-round-robin.js]:
```suggestion
			if (!match || match.state !== 'available') continue;
```

File: server/tournaments/generator-round-robin.js [hunk 22]
```diff
@@ -113,24 +113,24 @@ class RoundRobin {
...
 			match.score = [0, 1];
 			++this.userScores[col];
 			--this.totalPendingMatches;
-		});
+		}
 
-		this.matches.forEach((challenges, row) => {
+		for (const [row, challenges] of this.matches.entries()) {
 			let match = challenges[userIndex];
 			if (!match || match.state !== 'available') return;
```
Comment (Zarel) [inline server/tournaments/generator-round-robin.js]:
```suggestion
			if (!match || match.state !== 'available') continue;
```

File: server/tournaments/generator-round-robin.js [hunk 23]
```diff
@@ -155,14 +155,14 @@ class RoundRobin {
 
 		/** @type {[User, User][]} */
 		let matches = [];
-		this.matches.forEach((challenges, row) => {
-			challenges.forEach((match, col) => {
+		for (const [row, challenges] of this.matches.entries()) {
+			for (const [col, match] of challenges.entries()) {
 				if (!match) return;
```
Comment (Zarel) [inline server/tournaments/generator-round-robin.js]:
```suggestion
				if (!match) continue;
```

File: server/tournaments/index.js [hunk 24]
```diff
@@ -483,21 +482,21 @@ class Tournament {
...
 			challengeBys.get(match[1]).push(match[0]);
 
 			this.availableMatches.get(match[0]).set(match[1], true);
-		});
+		}
 
 		const now = Date.now();
-		this.availableMatches.forEach((availableMatches, user) => {
+		for (const [user, availableMatches] of this.availableMatches) {
 			if (oldAvailableMatches.get(user)) return;
```
Comment (Zarel) [inline server/tournaments/index.js]:
```suggestion
			if (oldAvailableMatches.get(user)) continue;
```

File: server/tournaments/index.js [hunk 25]
```diff
@@ -645,7 +644,7 @@ class Tournament {
 		}
 		if (this.autoDisqualifyTimer) clearTimeout(this.autoDisqualifyTimer);
 		const now = Date.now();
-		this.lastActionTimes.forEach((time, player) => {
+		for (const [player, time] of this.lastActionTimes) {
 			let availableMatches = false;
 			if (this.availableMatches.get(player).size) availableMatches = true;
 			const pendingChallenge = this.pendingChallenges.get(player);
```
Comment (Zarel) [inline server/tournaments/index.js]: There's a lot of returns in the lines below this that need to be changed to `continue`

Comment (Zarel) [reply server/tournaments/index.js]: (These are still not done)

Comment (TheJetOU) [reply server/tournaments/index.js]: https://github.com/Zarel/Pokemon-Showdown/pull/5269/files#diff-9d3af10d83c5fb18f33b23271bd2d3f0R654, they're done or did I miss one

File: server/punishments.js [hunk 26]
```diff
@@ -67,10 +67,12 @@ class PunishmentMap extends Map {
 	 * @param {(punishment: Punishment, id: string, map: PunishmentMap) => void} callback
 	 */
 	forEach(callback) {
-		super.forEach((punishment, k) => {
-			if (Date.now() < punishment[2]) return callback(punishment, k, this);
+		for (const [k, punishment] of super.entries()) {
+			if (Date.now() < punishment[2]) {
+				return callback(punishment, k, this);
```
Comment (Zarel) [inline server/punishments.js]: Why did you change this back? It's supposed to be `callback` and then `continue`

File: server/punishments.js [hunk 27]
```diff
@@ -287,17 +289,17 @@ Punishments.savePunishments = function () {
...
+		}
 		return buf;
 	});
 };
 
 Punishments.saveRoomPunishments = function () {
 	FS(ROOM_PUNISHMENT_FILE).writeUpdate(() => {
 		const saveTable = new Map();
-		Punishments.roomIps.nestedForEach((/** @type {Punishment} */ punishment, /** @type {string} */ roomid, /** @type {string} */ ip) => {
+		Punishments.roomIps.nestedforEach((/** @type {Punishment} */ punishment, /** @type {string} */ roomid, /** @type {string} */ ip) => {
```
Comment (Zarel) [inline server/punishments.js]:
```suggestion
		Punishments.roomIps.nestedForEach((/** @type {Punishment} */ punishment, /** @type {string} */ roomid, /** @type {string} */ ip) => {
```

Uppercase `F`; you also have this problem at every other instance of this.

File: server/punishments.js (hunk 26 above)
Comment (TheJetOU) [reply server/punishments.js]: Because, it's an eslint build error
https://eslint.org/docs/rules/callback-return

File: server/punishments.js [hunk 28]
```diff
@@ -404,9 +406,9 @@ Punishments.appendSharedIp = function (ip, note) {
 
 Punishments.saveSharedIps = function () {
 	let buf = 'IP\tType\tNote\r\n';
-	Punishments.sharedIps.forEach((/** @type {string} */ note, /** @type {string} */ ip) => {
+	for (const [/** @type {string} */note, /** @type {string} */ip] of Punishments.sharedIps) {
```
Comment (Zarel) [inline server/punishments.js]: No, this isn't a real each-function, you have to use `forEach`.

Comment (Zarel) [reply server/punishments.js]: You could also like rewrite the current `forEach` overrides as iterators, but that seems beyond you.

File: server/punishments.js (hunk 26 above)
Comment (Zarel) [reply server/punishments.js]: Just `eslint-disable` it.

Don't write bad code just because ESLint tells you to. :|

File: server/punishments.js [hunk 29]
```diff
@@ -136,12 +140,12 @@ class NestedPunishmentMap extends Map {
 	 * @param {(punishment: Punishment, roomid: string, userid: string) => void} callback
 	 */
 	nestedForEach(callback) {
-		this.forEach((subMap, k1) => {
-			subMap.forEach((punishment, k2) => {
+		for (const [k1, subMap] of this.entries()) {
+			for (const [k2, punishment] of subMap.entries()) {
 				if (Date.now() < punishment[2]) return callback(punishment, k1, k2);
```
Comment (Zarel) [inline server/punishments.js]:
```suggestion
				if (Date.now() < punishment[2]) {
					// eslint-disable-next-line callback-return
					callback(punishment, k1, k2);
					continue;
				}
```

File: server/tournaments/index.js (hunk 25 above)
Comment (Zarel) [reply server/tournaments/index.js]: You're right, they're done now.

**Comment by Zarel** [comment]
Thanks!

File: test/simulator/misc/decisions.js [hunk 30]
```diff
@@ -393,12 +401,20 @@ describe('Choices', function () {
 			]]);
 
 			battle.makeChoices('move lunardance, move lunardance', 'move lunardance, move lunardance');
-			battle.sides.forEach(side => side.active.forEach(pokemon => assert.fainted(pokemon)));
+			for (const side of battle.sides.values()) {
```
Comment (scheibo) [inline test/simulator/misc/decisions.js]: FYI this causes tests to fail for developers who are on Node versions older than 10.9. https://node.green/#ES2015-built-in-extensions-Array-prototype-methods-Array-prototype-values. https://github.com/Zarel/Pokemon-Showdown/pull/5282 to fix.
